src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        private class GeneratedFileReplacingCompilationTracker : ICompilationTracker
+        {
+            private readonly ICompilationTracker _underlyingTracker;
+            private readonly SourceGeneratedDocumentState _replacedGeneratedDocumentState;
+
+            private Compilation? _compilationWithReplacement;
```

Comment:
doc threading.

Reply:
do we need an interlocked read on it though to amke sure it looks consistently written to?

Reply:
So the method that ensures it's set does an interlocked write so that should be fine; it's just the regular "if null, then interlocked write and return" pattern used elsewhere. Earlier there were multiple fields initialized at once but actually now that there's just one I can simplify this.